### PR TITLE
Update Multistart API

### PR DESCRIPTION
**Pull request by edyounis**
Changed the instantiation API by moving the multi-start handling from `circuit.instantiate` to the `Instantiater`.

### CODE REVIEW

File: bqskit/ir/circuit.py [hunk 1]
```diff
@@ -2752,108 +2761,8 @@ def instantiate(
 
         instantiater = cast(Instantiater, instantiater)
 
-        # Check Target
-        try:
```
Comment (alonkukl) [inline bqskit/ir/circuit.py:2756]: Removing this casting is a big change, that breaks the API, are you sure you want to do that?

File: bqskit/ir/opt/instantiater.py [hunk 2]
```diff
@@ -54,6 +56,42 @@ def instantiate(
...
+            circuit (Circuit): The circuit template to instantiate.
+
+            target (UnitaryMatrix | StateVector | StateSystem): The unitary
+                matrix to implement or state to prepare.
+
+            num_starts (int): The number of starting points to attempt
+                instantiation with.
+
+        Returns:
+            (Circuit): The paremeters for the circuit that makes the
```
Comment (alonkukl) [inline bqskit/ir/opt/instantiater.py:78]: It should be the circuit with the best parameters

**Review by alonkukl — CHANGES_REQUESTED** [review]
I think that the API change of not converting the target is a big difference that should be talked over.

File: bqskit/ir/circuit.py (hunk 1 above)
Comment (edyounis) [reply bqskit/ir/circuit.py:2756]: I think the casting there was redundant as it was also cast inside Instantiater as well.

File: bqskit/ir/opt/instantiater.py [hunk 3]
```diff
@@ -54,6 +56,42 @@ def instantiate(
...
+            Instantiater object may happen in parallel.
+        """
+        target = self.check_target(target)
+        start_gen = RandomStartGenerator()
+        starts = start_gen.gen_starting_points(num_starts, circuit, target)
+        cost_fn = HilbertSchmidtCostGenerator().gen_cost(circuit, target)
+        params_list = [self.instantiate(circuit, target, x0) for x0 in starts]
+        params = sorted(params_list, key=lambda x: cost_fn(x))[0]
+        circuit.set_params(params)
+        return circuit
```
[the remark this answers is not in the document]
Comment (edyounis) [reply bqskit/ir/opt/instantiater.py:93]: We want to return the same circuit here. That way, `circuit.instantiate(...)` and `circuit = circuit.instantiate(...)` mean the same thing, which is especially helpful when mapping `instantiate` calls through the runtime. I added a note mentioning this.

[the remark this answers is not in the document]
Comment (alonkukl) [reply bqskit/ir/opt/instantiater.py:94]: It seems strange that the function returns and changes in place.

**Pull request merged**